### PR TITLE
Added the ability to stop (kill) the UrlRequest thread

### PR DESCRIPTION
**Pull request by misl6**
Now the user can stop the UrlRequest async download by killing the thread.
I've also implemented a on_kill callback.

### CODE REVIEW

File: kivy/network/urlrequest.py [hunk 1]
```diff
@@ -509,6 +514,14 @@ def _dispatch_result(self, dt):
                     if func:
                         func(self, data[0], data[1])
 
+            elif result == 'killed':
+                if self._debug:
+                    Logger.debug('UrlRequest killed by user')
```
Comment (pythonic64) [inline kivy/network/urlrequest.py]: You forgot `:` in log message. It should be:
`'UrlRequest: Killed by user'`

**Comment by pythonic64** [comment]
Maybe better wording would be to use `cancel` instead of `stop` and `on_cancel` instead of `on_kill`.

**Comment by misl6** [comment]
> Maybe better wording would be to use `cancel` instead of `stop` and `on_cancel` instead of `on_kill`.

Agree. Pushed a commit with the fixes you proposed.

File: kivy/network/urlrequest.py [hunk 2]
```diff
@@ -224,13 +228,7 @@ def run(self):
         q = self._queue.appendleft
         url = self.url
         req_body = self.req_body
-        req_headers = self.req_headers or {}
-        if (
-            Config.has_section('network')
-            and 'useragent' in Config.items('network')
-        ):
-            useragent = Config.get('network', 'useragent')
-            req_headers.setdefault('User-Agent', useragent)
```
[the remark this answers is not in the document]
Comment (misl6) [reply kivy/network/urlrequest.py]: Absolutely not. I don't know how it's happened. Just Re-added this part.

File: kivy/network/urlrequest.py [hunk 3]
```diff
@@ -175,10 +177,10 @@ class UrlRequest(Thread):
...
     def __init__(self, url, on_success=None, on_redirect=None,
                  on_failure=None, on_error=None, on_progress=None,
-                 req_body=None, req_headers=None, chunk_size=8192,
-                 timeout=None, method=None, decode=True, debug=False,
-                 file_path=None, ca_file=None, verify=True, proxy_host=None,
-                 proxy_port=None, proxy_headers=None, user_agent=None):
+                 on_cancel=None, req_body=None, req_headers=None,
+                 chunk_size=8192, timeout=None, method=None, decode=True,
+                 debug=False, file_path=None, ca_file=None, verify=True,
+                 proxy_host=None, proxy_port=None, proxy_headers=None):
```
[the remark this answers is not in the document]
Comment (misl6) [reply kivy/network/urlrequest.py]: Fixed.

File: kivy/network/urlrequest.py [hunk 4]
```diff
@@ -171,14 +170,17 @@ class UrlRequest(Thread):
...
+            operation via the .cancel() method.
     '''
 
     def __init__(self, url, on_success=None, on_redirect=None,
                  on_failure=None, on_error=None, on_progress=None,
                  req_body=None, req_headers=None, chunk_size=8192,
                  timeout=None, method=None, decode=True, debug=False,
                  file_path=None, ca_file=None, verify=True, proxy_host=None,
-                 proxy_port=None, proxy_headers=None, user_agent=None):
+                 proxy_port=None, proxy_headers=None, on_cancel=None):
```
Comment (tshirtman) [inline kivy/network/urlrequest.py]: this still removes the `user_agent` parameter, please add it back.

Comment (misl6) [reply kivy/network/urlrequest.py]: My bad. Fixed it.

**Pull request merged**